### ISSUE TITLE
Cisco ssh "LoginFailure: Login failed: No existing session" using max_thread>1

### ISSUE THREAD

**Issue by ninnux** [knipknap/exscript#107]
Script randomly fails to access into some cisco routers (2921,1941).
It raises LoginFailure: Login failed: No existing session using using queue parameters max_thread>1
Running same script many times failed routers change.
What kind of data can I give you for troubleshoot it?
Thank you
Nino

**Comment by egroeper** [comment]
You are running vanilla (unpatched) exscript version?
Which version are you using?
Are only these two device types affected?
The issue doesn't appear with max_thread=1?

The question is, if this is an issue with the underlying paramiko SSH library or with threading in exscript.
If it's a paramiko issue I would assume, that it would occur with max_thread=1, too. Perhaps you could try a simple single connection to an affected device with debugging enabled:

```
from Exscript.util.interact import read_login
from Exscript.protocols import SSH2

account = read_login()

for x in range(1,100):
    conn = SSH2(debug = 1)
    conn.connect('<affected host>')
    conn.login(account)
    conn.execute('<some simple command like show sys>')
    conn.send('exit\r')
    conn.close()
```

You can get more output when using Queue by increasing verbosity:
```
quickstart('<affected host>', do_something, verbose=2)
```

**Comment by ninnux** [comment]
I have no newline problem too, so i patched ios driver adding options on newline regexp, but issue is for first connection of day, so after first round I can use vanilla version.
So, Yes, I'm using master branch (446cf4d75a5e0bee2f37f56a8f704c68f52f91a5).
All cisco routers I tested (19xx, 29xx, 36xx) have this issue.
With max_thread=1 issue disappear.
My code use:
queue = Queue(verbose = 0, max_threads = 10)
queue.run(hosts, do_something)

increasing queue verbose parameter have same effect than quickstart one?

**Comment by egroeper** [comment]
Interesting. The exception comes from the underlying paramiko SSH library.
Which version of paramiko are you using?
Can you provide the complete Traceback of your exception?

**Comment by egroeper** [comment]
You are trying to authenticate using passwords? 
Do you have any active SSH key in your ssh-agent?

It seems like this issue is related with SSH keys quite often. But looking at the exscript code, it looks like this shouldn't happen, if password authentication is used.

**Comment by ninnux** [comment]
Paramiko version is: 1.10.1-1git1build1 (ubuntu package)
Yes, I'm trying authenticate using password.
Yes, I had some active kays on my ssh-agent, now I deactivate them but issue persists.
Traceback is:
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/workqueue/Job.py", line 64, in run
    self.function(self)
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/Queue.py", line 80, in _wrapped
    result = func(job, host, conn, *args, **kwargs)
  File "cisco-garr-mass-configs.py", line 17, in do_something
    conn.authenticate()
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/protocols/Protocol.py", line 650, in authenticate
    self.protocol_authenticate(account)
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/protocols/Protocol.py", line 676, in protocol_authenticate
    self._protocol_authenticate(user, password)
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/protocols/SSH2.py", line 266, in _protocol_authenticate
    self._paramiko_auth(user, password)
  File "/usr/local/lib/python2.7/dist-packages/Exscript-v2.1.503_g446cf4d-py2.7.egg/Exscript/protocols/SSH2.py", line 245, in _paramiko_auth
    raise LoginFailure('Login failed: ' + str(last_exception))
LoginFailure: Login failed: No existing session
```
Where did you found that is an issue related to paramiko and SSH keys?
Is a known issue?

**Comment by egroeper** [comment]
I found it [here](https://www.google.com/search?channel=fs&q=%22No+existing+session%22+paramiko).

I'm not sure, if this is still an issue. It's just a guess. The issues / articles are pretty old. But I know the exception comes from paramiko and there has to be some reason.
If the issue persists without having SSH keys activated, we need to have another look at it.

**Comment by ninnux** [comment]
Are you noted two different tracebacks?
One raise "No existing session" and the other one raise "Authentication failed". 
How can I help you? can I make some other tests?
Let me know

**Comment by iargue** [comment]
In case you (Or anyone else is experiencing this issue from google, since I found this issue several times in google).

I came across this issue of "No existing session" when trying to SSH into a Cisco Wireless Controller. Using 5 or more threads at once would cause "No existing session" failures. I didn't have any SSH keys or agents or anything at all, and so these extra old paramiko errors did not relate.

After several hundred hours of debugging, I found the problem, and the solution. During the debugging process, I moved everything to gevent, and so if you still have issues with the built in threading process, you'll need to dig deeper into the code.

The problem, as I found it, is that if you don't login (Or even send anything) to the controller within a small window of time, it closes the session. This only happens before login, and after login the SSH timeout value actually matters. So the issue that I was having is that it would take too long into log into a controller, and the controller would close the connection. When the script reached this controller, it would try and login, and paramiko would complain because the SSH connection was closed.

With threading in place, it didn't make any sense why it would take too login to do a simple login, and so I had to go back to the very basics, and start timing everything that I did, to find out where it was going wrong. In this process I moved over to Gevent as they do an excellent job of switching on blocking, and it still didn't fix the problem, and so I had to keep looking. Eventually, I found the problem to be with this line of code.

https://github.com/knipknap/exscript/blob/446cf4d75a5e0bee2f37f56a8f704c68f52f91a5/src/Exscript/protocols/Protocol.py#L624

with self._get_account(account) as account:

This is in protocol.py under def login. This little line right here forces every single login to be done one at a time, and doesn't allow (No matter what I did) for others to work while the login is happening. Each login was taking 1.8s for me, but 90% of it was waiting for data and nothing else, and so blocking while this happened was bad. To fix this, I changed that line to

account = account.context()

and it resolved my issue. I think this may only work if you create an account object and pass it to the login command, otherwise you might need to modify the fix. After doing this, I went from barely making it through 10 connections to connecting and logging into all 255 wireless controllers within 22 seconds.

I hope this fixes the problem for anyone else googling this. I spent far to many hours breaking apart the script in hopes of fixing the problem. At this point, I am still running this script as a gevent based thread instead of the built in threading of exscript, as Gevent patches the socket module so it will not block while waiting for data, and thus you can do more work at once. I think using the gevent patch is needed because each SSH connection took 5s to establish the connection (Happens the same if using putty or anything else. WLC is SLOW to response at the start). Using Gevent allowed me to start the connection for all controllers at once, otherwise it would be waiting 5s for every thread before ever trying to connect.

Even if this issue is years old. If you stumble across this post while frantically trying to get python and cisco devices to play friendly, please feel free to message me. I work and program with cisco devices every day, and likely found and fixed things you are having a problem with (Or at the least, am interested in helping find the problem)

**Comment by knipknap** [comment]
This is very interesting. _get_account actually locks for a very good
reason: Authentication servers often assign a sequence id to a login
attempt, similar to session ids on websites, and then use the id to
identify a session.

So we had to make sure that threads don't "steal" each others sequence
numbers during login. Of course, this is only needed if multiple devices
share a common authentication server, otherwise it has a drastic impact on
parallelisation. Perhaps there should be a way to make this locking
optional.

-Samuel

**Comment by iargue** [comment]
I'm a little lost as to how this would create issues?

I logged into all 255 WLC's in 22s without any session ID issues or anything at all and they are all on Cisco Tacacs (So maybe if a radius server or something handles it poorly?).

There shouldn't be an issue for password logins (Maybe for other logins) as our scripts don't do session id's or really anything for the authentication. On our end, we just send it a username and password, and its the end device that does the checking with the authorization severs and not us, and so session id's don't matter to us at all. The session id's are there specifically to allow us to do exactly what we want to do here, login to multiple devices at once without creating errors or stealing someone else's session id.

If you could provide me with a use case where this is actually an error, I would love to see it.

**Comment by knipknap** [comment]
The issue depends on the authentication method. In particular, TACACS
supports OTP (S/KEY) authentication. Essentially, it makes the login
procedure look something like this:

    Welcome to this Cisco Router!
    Username: foobar
    Please enter the s/key password 1234 and salt it using the string
"ASDF": *****
    router>

So when two threads enter the username simultaneously, they are both asked
for s/key password 1234. The first thread to deliver the correct password
causes this sequence number to be invalidated, causing the other thread to
fail. To avoid this, the account needs to be locked at the time where the
username is entered, and can be unlocked as soon as the password was
confirmed.

**Comment by iargue** [comment]
There is absolutely a method to handle this thread safe, without the need for blocking, or significantly less blocking. I will have to fork the environment and test authentication under every supported authentication method to see how the current system handles each, and how non blocking would handle each to determine the fail cases.

i have two questions though,

First is where does exscript handle OTP? I see there is an external module for OTP but I don't see anywhere its used to generate a password and login to a system.

The second is what methods of authentication does exscript currently support? I only see password, ssh key, interactive, and agent. Are there more method of authentication somewhere else that I need to account for and test for?

I'll build a devolopment network this week that includes all forms of authentication supported on tacacs+/cisco devices.

**Comment by knipknap** [comment]
OTP is actually just an application layer add-on to the password
authentication method. On the protocol level, there is no difference - it
can only be handled by checking whether the password prompt requests an
s/key password. This is handled here:

https://github.com/knipknap/exscript/blob/master/src/Exscript/protocols/Protocol.py#L748

**Comment by knipknap** [comment]
I have commited a possible fix in 414724b3ad5d0cd57d565348e8c528aae551a2a5. It works by making the locking optional:

account = Account ('user', 'password', needs_lock=False)

Please let me know if the problem persists when setting needs_lock to False.

**Comment by knipknap** [comment]
I am closing this bug on the assumption that it works; feel free to re-open if the problem persists with needs_lock=False
